Description: Add loopback file support and loopoffset for ONIE/SONiC installer
 This is a refreshed combination of two patches originally written for
 Debian 0.142, now adapted for Ubuntu 0.151ubuntu1:
  - loopback-file-system-support (basic LOOP variable exports, mount_loop_root,
    NFS loopback, readonly condition with ntfs check)
  - loopback-file-offset-support (LOOPOFFSET variable, losetup)
 .
 Ubuntu 0.151ubuntu1 already has significant loopback support inlined in
 scripts/local (loop=, loopflags=, loopfstype= cmdline parsing + inlined
 mounting logic), but is missing: exports for LOOP* variables, LOOPOFFSET
 support, NFS loopback, and the ntfs-aware readonly condition.
 .
 Forward-ported by Sheldon Qi <sheldon.qi@canonical.com> 2026-07-24.
Origin: https://bugs.debian.org/cgi-bin/bugreport.cgi?att=0;bug=468114;msg=32
Bug-Debian: https://bugs.debian.org/468114
Last-Update: 2026-07-24
---
 init              |    8 ++++++++
 initramfs-tools.7 |    5 +++++
 scripts/functions |   48 ++++++++++++++++++++++++++++++++++++++++++++++++
 scripts/local     |   20 +++++++++++++++++---
 scripts/nfs       |    4 ++++
 5 files changed, 82 insertions(+), 3 deletions(-)

--- initramfs-tools-0.151ubuntu1.orig/init	2026-04-02 10:33:36.000000000 +0000
+++ initramfs-tools-0.151ubuntu1/init	2026-07-24 15:16:59.402560571 +0000
@@ -60,6 +60,10 @@
 export ROOTDELAY=
 export ROOTFLAGS=
 export ROOTFSTYPE=
+export LOOP=
+export LOOPFLAGS=
+export LOOPFSTYPE=
+export LOOPOFFSET=
 export IP=
 export IP6=
 export VLAN=
@@ -146,6 +150,10 @@
 		# shellcheck disable=SC2034
 		LOOPFSTYPE="${x#loopfstype=}"
 		;;
+		loopoffset=*)
+			# shellcheck disable=SC2034
+			LOOPOFFSET="${x#loopoffset=}"
+			;;
 	nfsroot=*)
 		# shellcheck disable=SC2034
 		NFSROOT="${x#nfsroot=}"
--- initramfs-tools-0.151ubuntu1.orig/initramfs-tools.7	2026-04-02 09:05:04.000000000 +0000
+++ initramfs-tools-0.151ubuntu1/initramfs-tools.7	2026-07-24 15:18:02.674875868 +0000
@@ -71,6 +71,11 @@
 set the loop file system mount option string, if applicable.
 
 .TP
+\fB\fI loopoffset
+set the loop file offset from which to mount the loop, if applicable.
+The default is 0 and requires loop to be defined.
+
+.TP
 \fB\fI nfsroot
 can be either "auto" to try to get the relevant information from DHCP or a
 string of the form NFSSERVER:NFSPATH or NFSSERVER:NFSPATH:NFSOPTS.
--- initramfs-tools-0.151ubuntu1.orig/scripts/functions	2026-04-02 10:34:24.000000000 +0000
+++ initramfs-tools-0.151ubuntu1/scripts/functions	2026-07-24 15:17:05.601523555 +0000
@@ -1043,6 +1043,54 @@
 	echo "$result"
 }
 
+# Mount a loopback device, which is present on the mounted filesystem.
+mount_loop_root()
+{
+	mkdir -p /host
+	mount -o move "${rootmnt}" /host
+	loopfile="/host/${LOOP#/}"
+
+	while [ ! -e "$loopfile" ]; do
+		panic "ALERT! $loopfile does not exist.  Dropping to a shell!"
+	done
+
+	if [ "${readonly?}" = "y" ]; then
+		roflag=-r
+	else
+		roflag=-w
+	fi
+
+	# Get the loop filesystem type if not set
+	if [ -z "${LOOPFSTYPE}" ]; then
+	    FSTYPE=$(get_fstype "$loopfile")
+	else
+	    FSTYPE=${LOOPFSTYPE}
+	fi
+
+	# FIXME This has no error checking
+	modprobe loop
+	modprobe "${FSTYPE}"
+
+	if [ -n "${LOOPOFFSET}" ]; then
+		# create a loop device for the fs within the file
+		loopdev="$(losetup -f)"
+		losetup -o "${LOOPOFFSET:-0}" "${loopdev}" "${loopfile}" || \
+		    panic "ALERT! $loopdev could not be setup using $loopfile"
+		loopfile="$loopdev"
+	fi
+
+	# FIXME This has no error checking
+	if [ -z "${LOOPFLAGS}" ]; then
+		mount ${roflag} -o loop -t "${FSTYPE}" "$loopfile" "${rootmnt}"
+	else
+		mount ${roflag} -o loop -t "${FSTYPE}" "${LOOPFLAGS}" "$loopfile" "${rootmnt}"
+	fi
+
+	if [ -d "${rootmnt}/host" ]; then
+		mount -o move /host "${rootmnt}/host"
+	fi
+}
+
 # Mount the root file system.  It should be overridden by all
 # boot scripts.
 mountroot()
--- initramfs-tools-0.151ubuntu1.orig/scripts/local	2026-04-02 09:05:04.000000000 +0000
+++ initramfs-tools-0.151ubuntu1/scripts/local	2026-07-24 15:17:38.985960842 +0000
@@ -206,7 +206,7 @@
 	local_premount
 
 	if [ "${readonly?}" = "y" ] && \
-	   [ -z "$LOOP" ]; then
+	   { [ -z "$LOOP" ] || [ "${FSTYPE#ntfs}" = "$FSTYPE" ]; }; then
 		roflag=-r
 	else
 		roflag=-w
@@ -258,9 +258,21 @@
 		# FIXME This has no error checking
 		modprobe loop
 
+		if [ -n "${LOOPOFFSET}" ]; then
+			# create a loop device for the fs within the file
+			loopdev="$(losetup -f)"
+			losetup -o "${LOOPOFFSET:-0}" "${loopdev}" "/host/${LOOP#/}" || \
+			    panic "ALERT! $loopdev could not be setup using /host/${LOOP#/}"
+			LOOPFILE_IS_DEV=y
+		fi
+
 		# FIXME This has no error checking
 		# shellcheck disable=SC2086
-		mount ${roflag} -o loop -t ${FSTYPE} ${LOOPFLAGS} "/host/${LOOP#/}" "${rootmnt?}"
+		if [ "${LOOPFILE_IS_DEV}" = "y" ]; then
+			mount ${roflag} -t ${FSTYPE} ${LOOPFLAGS} "${loopdev}" "${rootmnt?}"
+		else
+			mount ${roflag} -o loop -t ${FSTYPE} ${LOOPFLAGS} "/host/${LOOP#/}" "${rootmnt?}"
+		fi
 
 		if [ -d "$rootmnt/host" ]; then
 			mount -o move /host "$rootmnt/host"
--- initramfs-tools-0.151ubuntu1.orig/scripts/nfs	2025-01-28 16:51:50.000000000 +0000
+++ initramfs-tools-0.151ubuntu1/scripts/nfs	2026-07-24 15:17:51.102335577 +0000
@@ -74,6 +74,10 @@
 
 	# shellcheck disable=SC2086
 	nfsmount -o nolock ${roflag} ${NFSOPTS} "${NFSROOT}" "${rootmnt?}"
+
+	if [ "${LOOP}" ]; then
+		mount_loop_root
+	fi
 }
 
 # NFS root mounting
